tactics/btermdn.ml

```diff
@@ -115,32 +213,136 @@ let constr_val_discr env sigma ts t =
       Label(GRLabel (ConstructRef cstr_sp), stack)
     | Var id when evaluable_named id env ts -> Everything
     | Var id -> Label(GRLabel (VarRef id), stack)
-    | Prod (n,d,c) -> Label(ProdLabel, [d; c])
-    | Lambda _ when Option.is_empty ts && List.is_empty stack -> Nothing
+    | Prod (n,d,c) -> Label(ProdLabel, [Constr d; Constr c])
+    | Lambda (_,d,c) when List.is_empty stack ->
+      decomp_lambda_constr env sigma ts decomp d c
     | Lambda _ -> Everything
     | Sort _ -> Label(SortLabel, [])
     | Evar _ -> Everything
     | Case (_, _, _, _, _, c, _) ->
       begin
         match decomp stack c with
         | Label (GRLabel (ConstructRef _), _) -> Everything (* over-approximating w.r.t. [fMATCH] *)
-        | Label _  | Nothing -> Label(CaseLabel, c :: stack)
+        | (Label _ | Nothing) as res -> Label(CaseLabel, PartialConstr res :: stack)
         | Everything -> Everything
       end
     | Rel _ | Meta _ | LetIn _ | Fix _ | CoFix _
     | Int _ | Float _ | String _ | Array _ -> Nothing
+    in
+    debug Pp.(fun () -> str "constr_val_discr.decomp output: " ++ pr_constr_res (Printer.pr_leconstr_env env Evd.empty) out);
```

Comment:
Shouldn't you pass the correct evarmap to the printer? We have it at hand already.

Reply:
I fixed the two places where we have `sigma`. Let me know if the other places can be/need to be fixed as well.